File: GeoQuizJava/app/src/main/java/com/akimov/android/geoquiz/CheatActivity.java
package com.akimov.android.geoquiz;

import android.content.Context;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import junit.framework.Test;

public class CheatActivity extends AppCompatActivity {
    private static final String EXTRA_ANSWER_IS_TRUE = "com.akimov.android.geoquiz.answer_is_true";
    private static final String EXTRA_ANSWER_IS_SHOWN = "com.akimov.android.geoquiz.answer_shown";
    private static final String TAG = "CheatActivity";

    private static final String KEY_ANSWERISSHOWN = "answer";

    private boolean isAnswerShown;
    private boolean mAnswerIsTrue;
    private Button mShowAnswerButton;
    private TextView mAnswerTextView;

    public static Intent newIntent(Context packageContext, boolean answerIsTrue) {
        Intent i = new Intent(packageContext, CheatActivity.class);
        i.putExtra(EXTRA_ANSWER_IS_TRUE, answerIsTrue);
        return i;
    }

    public static boolean wasAnswerShown(Intent result) {
        return result.getBooleanExtra(EXTRA_ANSWER_IS_SHOWN, false);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cheat);

        mAnswerIsTrue = getIntent().getBooleanExtra(EXTRA_ANSWER_IS_TRUE, false);
        mAnswerTextView = (TextView) findViewById(R.id.answer_text_view);

        if (savedInstanceState != null) {
            isAnswerShown = savedInstanceState.getBoolean(KEY_ANSWERISSHOWN);
            if (isAnswerShown) {
                setAnswerShownResult();
                if (mAnswerIsTrue)
                    mAnswerTextView.setText(R.string.true_button);
                else
                    mAnswerTextView.setText(R.string.false_button);
            }
        }


        mShowAnswerButton = (Button) findViewById(R.id.show_answer_button);
        mShowAnswerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mAnswerIsTrue)
                    mAnswerTextView.setText(R.string.true_button);
                else
                    mAnswerTextView.setText(R.string.false_button);

                isAnswerShown = true;
                setAnswerShownResult();
            }
        });


    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        Log.i(TAG, "onSaveInstanceState");
        outState.putBoolean(KEY_ANSWERISSHOWN, isAnswerShown);
        super.onSaveInstanceState(outState);
    }

    public void setAnswerShownResult() {
        Intent data = new Intent();
        data.putExtra(EXTRA_ANSWER_IS_SHOWN, isAnswerShown);
        setResult(RESULT_OK, data);
    }
}

File: GeoQuiz/app/src/main/java/com/akimov/geoquiz/presentation/cheat/view/CheatActivity.kt
package com.akimov.geoquiz.presentation.cheat.view

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import com.akimov.geoquiz.R
import kotlinx.android.synthetic.main.activity_cheat.*

class CheatActivity : AppCompatActivity() {

  private var mAnswerIsTrue: Boolean = false

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_cheat)

    mAnswerIsTrue = intent.getBooleanExtra(EXTRA_ANSWER_IS_TRUE, false)
  }

  fun onShowAnswerBtnClicked(view: View) {
    answer_textView.setText(if (mAnswerIsTrue) R.string.true_button else R.string.false_button)
    setAnswerShowResult(true)
  }

  private fun setAnswerShowResult(isAnswerShown: Boolean) {
    val data = Intent()
    data.putExtra(EXTRA_ANSWER_SHOWN, isAnswerShown)
    setResult(Activity.RESULT_OK, data)
  }

  companion object {
    private val EXTRA_ANSWER_IS_TRUE = "com.akimov.geoquiz.answer_is_true" //static final
    private val EXTRA_ANSWER_SHOWN = "com.akimov.geoquiz.answer_shown"
    //private val INTENT_
    fun newIntent(context: Context, answerIsTrue: Boolean): Intent {
      val intent = Intent(context, CheatActivity::class.java)
      intent.putExtra(EXTRA_ANSWER_IS_TRUE, answerIsTrue)
      return intent
    }

    fun wasAnswerShown(result: Intent): Boolean {
      return result.getBooleanExtra(EXTRA_ANSWER_SHOWN, false)
    }
  }
}
